numpy/lib/nanfunctions.py

```diff
@@ -640,8 +641,15 @@ def nansum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
     nan
 
     """
-    a, mask = _replace_nan(a, 0)
-    return np.sum(a, axis=axis, dtype=dtype, out=out, keepdims=keepdims)
+    try:
+        a = np.asanyarray(a)
```

Comment:
Is this `asanyarray` necessary for any test? I would have thought `np.sum` would take care.

Reply:
As I noted on the mailing list, this should probably be `np.asarray(a)` instead, which would obviate the need for the fallback logic.

Reply:
Yes, that would avoid the `matrix` problems and keep the behaviour for duck arrays the same, but break other subclasses such as `MaskedArray` and `Quantity` (which work fine with the present implementation), and will not allow duck arrays to override by updating their `sum` method or correctly implementing `__array_ufunc__`.

But what would work is to use your suggestion for the fall-back logic:
```
try:
    return np.sum(a, axis=axis, dtype=dtype, out=out, keepdims=keepdims,
                  where=_nan_mask(a))
except TypeError:
    return np.asarray(a).sum(....)
```

Reply:
Actually, as noted in the thread, the `asarray` call is sadly not OK as we currently use `asanyarray` - and it thus won't solve the fact that `matrix` cannot handle it.